doc/BMI_MODELS.md

```diff
@@ -0,0 +1,151 @@
+  * this is needed in here so the realization can directly access things implicit in the file, like times, time step amounts, and time step sizes
+* `uses_forcing_file`
+  * boolean indicating whether the backing BMI model is written to read input data from the forcing file (as opposed to receiving it via getters)
+* `init_config` 
+  * the string path to the BMI initialization config file for the catchment
+* `main_output_variable` 
+  * the string value of the primary output variable
+  * this is the value that returned by the realization's `get_response()`
+  * the string must match an item return by the relevant variant of the BMI `get_output_var_names()` function
+  
+### Optional Parameters
+* `other_input_variables`
+  * this may be provided to set certain model variables more directly after its `initialize` function
+  * JSON structure should be one or more nested JSON nodes, keyed by the variable name, and with the variable values contained as a list
+  * e.g.,  `"other_input_variables": {"ex_var_1": [0, 1, 2]}`
+* `output_variables`
+  * can specify the particular set and order of output variables to include in the realization's `get_output_line_for_timestep()` (and similar) function
+  * JSON structure should be a list of strings
+  * if not present, defaults to whatever it returned by the model's BMI `get_output_var_names()` function *the first time* it is invoked
+* `output_header_fields`
+  * can specify the header strings to use for the realization's printed output (i.e., the value returned by `get_output_header_line()`)
+  * JSON structure should be a list of strings
+  * when not present, the literal variable names are used
+  * when present, does not do any checking for ordering/correspondence compared to the output ordering of the variable values, so users must take care that ordering is consistent
+* `allow_exceed_end_time`
+  * boolean value to specify whether a model is allowed to execute `Update` calls that go beyond its end time (or the max forcing data entry)
+  * implied to be `false` by default
+  
+## BMI Models Written in C
+
+* [Model As Shared Library](#bmi-c-model-as-shared-library)
+* [Example](#bmi-c-cfe-example)
+* [Caveats](#bmi-c-caveats)
+
+### BMI C Model As Shared Library
+
+For **C** models, the model must be packaged as a pre-compiled shared library.  Several CMake cache variables must be configure for controlling whether to expect such a library and how to find it:
+
+* `BMI_C_LIB_ACTIVE` 
+  * type: `BOOL` 
+  * must be set to `ON` (or equivalent in CMake) for BMI C shared library functionality to be compiled and active
+* `BMI_C_LIB_NAME`
+  * type: `STRING `
+  * must be set if `BMI_C_LIB_ACTIVE` is `ON` to supply the appropriate name for the shared library
+* `BMI_C_LIB_HINT_DIR`
+  * type: `STRING`
+  * may be set to provide a [hint](https://cmake.org/cmake/help/latest/command/find_library.html) to CMake when it tries to find the library
```

Comment:
I was trying to create separate, isolated control mechanisms in the variables, but I also like to over complicate ...

Reply:
In all seriousness, yes, `CMAKE_PREFIX_PATH` would work, as would `CMAKE_LIBRARY_PATH` or `CMAKE_FRAMEWORK_PATH`.  But I'm not sure if this then could impact behavior beyond the BMI C configuration.  In practice, it'd have to be in a pretty contrived situation, but I liked the idea of dedicated controls, as well as somewhat self-documenting config variables.

Reply:
It is not uncommon, for sure.  It might be worth putting together a cmake config file example for the external, then we could also use a `find_package`, that is actually where CMAKE_PREFIX_PATH came to mind.  Then the config could expose the lib dirs.  I would remove the "hint" though, cause it isn't a hint so much as a concrete path that has to exist to find the lib to link?